chain/events/filter/event.go

```diff
@@ -27,22 +27,32 @@ func isIndexedValue(b uint8) bool {
 	return b&(types.EventFlagIndexedKey|types.EventFlagIndexedValue) > 0
 }
 
-type EventFilter struct {
-	id         types.FilterID
-	minHeight  abi.ChainEpoch // minimum epoch to apply filter or -1 if no minimum
-	maxHeight  abi.ChainEpoch // maximum epoch to apply filter or -1 if no maximum
-	tipsetCid  cid.Cid
-	addresses  []address.Address   // list of f4 actor addresses that are extpected to emit the event
-	keys       map[string][][]byte // map of key names to a list of alternate values that may match
-	maxResults int                 // maximum number of results to collect, 0 is unlimited
+type AddressResolver func(context.Context, abi.ActorID, *types.TipSet) (address.Address, bool)
+
+type EventFilter interface {
+	Filter
+
+	TakeCollectedEvents(context.Context) []*CollectedEvent
+	CollectEvents(context.Context, *TipSetEvents, bool, AddressResolver) error
+}
+
+type eventFilter struct {
+	id        types.FilterID
+	minHeight abi.ChainEpoch // minimum epoch to apply filter or -1 if no minimum
+	maxHeight abi.ChainEpoch // maximum epoch to apply filter or -1 if no maximum
+	tipsetCid cid.Cid
```

Comment:
can you remind me how can one get TipsetCid? (which is different from tipset keys CID?

Reply:
`TipSetKey#Cid()` does it, bundles up the raw bytes of each CID, concatenates them, and makes a CID from those bytes.

This internal API is serving the existing Eth filter API which can take an Eth `BlockHash`, in which case it converts that to a tipset CID from the internal index, and it's now serving this new API which takes a `TipSetKey` which it does a `.Cid()` on to get the CID.

wew .. that's not confusing at all!